### PR TITLE
Drand test driving

### PR DESCRIPTION
**Pull request by willscott**
* [x] refactor the demo `node` to an interface
* [x] write an implementation of the interface that runs 'in-process', spinning up a drand daemon thread, but not using commands.

### CODE REVIEW

**Comment by nikkolasg** [comment]
I would like to still being able to use the commands tbh, as the main tests are not testing enough or said differently, the test integration is meant to be sure that running a network using the CLI works.

**Comment by willscott** [comment]
> I would like to still being able to use the commands tbh, as the main tests are not testing enough or said differently, the test integration is meant to be sure that running a network using the CLI works.

Agreed. This first part of the change makes an interface over the CLI commands that are being dispatched by the orchestrator in the demo. I'm now re-implementing that interface with direct signalling to the daemon as well. Success will be when the demo test can run in both modes :) The in-process variant will work a bit better for testground, since it proves some additional opportunities for instrumenting each running instance.

File: demo/demo_test.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package main
+
+import "testing"
+
+func TestLocalOrchestration(t *testing.T) {
```
Comment (nikkolasg) [inline demo/demo_test.go:5]: This doesn't stop the whole network and doesn't do a resharing correct ?

File: demo/node/node_inprocess.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		l.log.Error("drand", "can't get cokey", "err", err)
+		return false
+	}
+	sdist := hex.EncodeToString(key.GetCoKey())
+	fmt.Printf("\t- Node %s has cokey %s\n", l.PrivateAddr(), sdist[10:14])
+	return true
+}
+
+func (l *LocalNode) Ping() bool {
+	cl, err := net.NewControlClient(l.ctrlAddr)
```
Comment (nikkolasg) [inline demo/node/node_inprocess.go:216]: By creating the control client once (so the connection is performed once) we might get some performance benefit _on circleCI_ :)

File: demo/node/node_inprocess.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+	}
+	if err := cl.Ping(); err != nil {
+		l.log.Error("drand", "can't ping", "err", err)
+		return false
+	}
+	return true
+}
+
+func (l *LocalNode) GetBeacon(groupPath string, round uint64) (resp *drand.PublicRandResponse, cmd string) {
+	c := core.NewGrpcClient()
```
Comment (nikkolasg) [inline demo/node/node_inprocess.go:229]: Same for here - creating the client once would probably avoid some RTT time, and anything is costly on circleCI

File: test/test.go [hunk 4]
```diff
@@ -50,6 +50,45 @@ func Addresses(n int) []string {
...
+func LocalHost() string {
+	addr, err := n.ResolveTCPAddr("tcp", "localhost:0")
+	if err != nil {
+		panic(err)
+	}
+	return addr.IP.String()
+}
+
+// FreeBind provides an address for binding on provided address
+func FreeBind(a string) string {
```
Comment (nikkolasg) [inline test/test.go:63]: This method was already somewhere before no ? Why copy it ?

File: demo/demo_test.go (hunk 1 above)
Comment (nikkolasg) [reply demo/demo_test.go:5]: I understand the demo with processes still test that - I'm just raising the point if we should do it there or not as well - could increase coverage as well.

Comment (willscott) [reply demo/demo_test.go:5]: I wanted to start with this and not delay unit tests more. I think this largely tests that the driver works enough to drive the local node, which is the main thing i want to not break as subsequent changes occur to the code base.

File: test/test.go (hunk 4 above)
Comment (willscott) [reply test/test.go:63]: we have free port. this one is more general in that it allows specification and returns the host part. so e.g. I can start with 'localhost' and learn if that translated into '127.0.0.1' or '[::1]', or i can start with '0.0.0.0' and learn that that translated into an external IP that i can then make a ssl cert for.

Is it worth finding all the uses of `FreePort` and having it either use this or refactoring the client to use this instead?

File: demo/demo_test.go (hunk 1 above)
Comment (willscott) [reply demo/demo_test.go:5]: for reference: this test takes less than a minute in CI, compared to ~3 minutes for the demo run

**Pull request merged**